source/Libraries/SteamLibrary/SteamShared/SteamTagNamer.cs

```diff
@@ -46,48 +47,88 @@ private string GetPackedTagNameFilePath()
             {
                 logger.Trace($"No new tag names found for '{settings?.LanguageKey}'");
                 fileName = GetPackedTagNameFilePath();
-            }
 
-            if (!File.Exists(fileName))
-            {
-                logger.Warn($"No tag names found for language '{settings?.LanguageKey}'");
-                return new Dictionary<int, string>();
+                if (!File.Exists(fileName))
+                {
+                    logger.Warn($"No tag names found for language '{settings?.LanguageKey}'");
+                    return new Dictionary<int, string>();
+                }
             }
+            var content = GetFileContents(fileName);
+
+            return content.response.tags.ToDictionary(x => x.tagid, x => x.name);
+        }
 
-            var tagNames = Serialization.FromJsonFile<Dictionary<int, string>>(fileName);
-            return tagNames;
+        private SteamTagFile GetFileContents()
+        {
+            return GetFileContents(GetTagNameFilePath()) ?? GetFileContents(GetPackedTagNameFilePath());
         }
 
-        private static Regex tagNameRegex = new Regex(@"data-param=""tags"" data-value=""(?<id>[0-9]+)"" data-loc=""(?<name>.+?)""", RegexOptions.Compiled);
+        private SteamTagFile GetFileContents(string path)
+        {
+            if (!File.Exists(path))
+                return null;
+
+            logger.Debug("Opening " + path);
+            try
+            {
+                var content = File.ReadAllText(path, Encoding.UTF8);
+                return Serialization.FromJson<SteamTagFile>(content);
+            }
+            catch (Exception e)
+            {
+                logger.Error(e, "Failed to read " + path);
+                throw;
+            }
+        }
 
         public Dictionary<int, string> UpdateAndGetTagNames()
         {
-            string url = $"https://store.steampowered.com/search/?l={settings?.LanguageKey}";
+            var existingFile = GetFileContents();
+
+            string url = $"https://api.steampowered.com/IStoreService/GetTagList/v1?language={settings?.LanguageKey}&have_version_hash={existingFile?.response?.version_hash}";
+
             logger.Debug("Downloading " + url);
-            var content = downloader.DownloadString(url);
-            var matches = tagNameRegex.Matches(content);
+            var httpClient = new HttpClient();
+            var downloadTask = httpClient.GetAsync(url);
+            downloadTask.Wait();
+            var downloadResult = downloadTask.Result;
+            var contentTask = downloadResult.Content.ReadAsStringAsync();
+            contentTask.Wait();
+
+            var filePath = GetTagNameFilePath();
 
-            var output = new Dictionary<int, string>();
-            foreach (var match in matches.Cast<Match>())
+            if (downloadResult.StatusCode != System.Net.HttpStatusCode.NotModified && !string.IsNullOrWhiteSpace(contentTask.Result))
```

Comment:
I'm not sure if this is necessary if you already downloaded all that data anyways.

Reply:
The response body is empty when it's NotModified, if that's what you mean